operator/ingress/v1/types.go

```diff
@@ -0,0 +1,110 @@
+package v1
+
+import (
+	configv1 "github.com/openshift/api/config/v1"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+
+// DNSRecord is a DNS record managed in the zones defined by
+// dns.config.openshift.io/cluster .spec.publicZone and .spec.privateZone.
+//
+// Cluster admin manipulation of this resource is not supported. This resource
+// is only for internal communication of OpenShift operators.
+type DNSRecord struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// spec is the specification of the desired behavior of the dnsecord.
+	Spec DNSRecordSpec `json:"spec,omitempty"`
+	// status is the most recently observed status of the dnsRecord.
+	Status DNSRecordStatus `json:"status,omitempty"`
+}
+
+// dnsRecordSpec are the details of a DNS record.
+type DNSRecordSpec struct {
+	// dnsName is the hostname of the DNS record
+	//
+	// +kubebuilder:validation:Required
+	// +required
+	DNSName string `json:"dnsName,omitempty"`
+	// targets are DNS resource record targets.
+	//
+	// +kubebuilder:validation:Required
+	// +required
+	Targets []string `json:"targets,omitempty"`
+	// recordType is a resource record type.
+	//
+	// The currently supported values are A and CNAME.
+	//
+	// +kubebuilder:validation:Required
+	// +required
+	RecordType DNSRecordType `json:"recordType,omitempty"`
+	// ttl is the resource record TTL in seconds. The default is 30.
+	//
+	// +kubebuilder:validation:Optional
+	// +optional
+	RecordTTL int64 `json:"recordTTL,omitempty"`
+}
+
+// dnsRecordStatus is the more recently observed status of the record.
+type DNSRecordStatus struct {
+	// zones are the status of the record per zone.
+	//
+	// +kubebuilder:validation:Optional
+	// +optional
+	Zones []DNSZoneStatus `json:"zones,omitempty"`
+}
+
+// dnsZoneStatus is the status of a record within a specific zone.
+type DNSZoneStatus struct {
+	// dnsZone is the zone where the record is expected to be managed.
+	//
+	// +kubebuilder:validation:Required
+	// +required
+	DNSZone configv1.DNSZone `json:"dnsZone"`
+	// conditions are any conditions associated with the record in the zone.
+	//
+	// +kubebuilder:validation:Optional
+	// +optional
+	Conditions []DNSZoneCondition `json:"conditions,omitempty"`
+}
+
+var (
+	// Failed means the record is not available within a zone.
+	DNSRecordFailedConditionType = "Failed"
+)
+
+// dnsZoneCondition is just the standard condition fields.
```

Comment:
`dnsZoneCondition` → `DNSZoneCondition`.

Reply:
Hold on, I thought godoc for types didn't go in the CRD, but this one _is_ in the CRD description—did CRD generation change?

Reply:
@damemi, is this a defect in the CRD generation?  It makes no sense to use the type name in this context: 
 https://github.com/openshift/api/blob/41c658872c442d2a9218fe9fa5e4885c260cc3ff/operator/ingress/v1/0000_50_dns-record.yaml#L70-L71

Reply:
Godoc for types does go in the CRD, it's where descriptions have always come from

Reply:
Hm, I see the same thing in other CRDs with lists of objects that have godoc (for example, I see the same issue with `spec.servers` in the `dnses.operator` CRD).

To be clear, the problem is that we define a type with godoc:
```go
// Foo is ...
type Foo {
  // ...
```
We define some type that we use in a CRD and that has a field for which the type of a slice of the first type:
```go
type Bar {
  // baz is a list of ...
  baz []Foo `json:"baz"`
  // ...
```
And the generated CRD description uses the first type's godoc, something like the following:
```yaml
spec:
  validation:
    openAPIV3Schema:
      type: object
      properties:
        type: object
        baz:
          description: baz is a list of ...
          type: array
          items:
            description: Foo is ...
```
In the context of the CRD, the type name is meaningless.

Assuming the CRD generation is working as intended, we should change our godoc to omit the type name if the type is used in a slice type for a field that is used in a CRD (in particular, for `dnsrecords`, that's `DNSZoneStatus` and `DNSZoneCondition`); agreed?

So we should change "dnsZoneCondition is just the standard condition fields." to "The standard condition fields." and "DNSZoneStatus is the status of a record within a specific zone." to "The status of a record within a specific zone."

Reply:
Isn't the convention with godoc to start a description with the type name? So that outcome might just be a requirement

Maybe instead, our comments should be more descriptive of the types they are using. Following your example, the doc could be `bas is a list of Foo`. So the user asks, what is Foo? And the next description tells them `Foo is ...`

Reply:
> Isn't the convention with godoc to start a description with the type name? So that outcome might just be a requirement

We already eschew that convention for fields.

> Maybe instead, our comments should be more descriptive of the types they are using. Following your example, the doc could be `bas is a list of Foo`. So the user asks, what is Foo? And the next description tells them `Foo is ...`

That would be weird and verbose but at least the result would be coherent.

Reply:
I think I got lost as to what further changes you guys want me to make. I'd like to get this merged today.

Reply:
The answer to all of these godoc questions seems unambiguous to me: the Go types should adhere to the godoc conventions, full stop. If the generator pipeline needs to do weird stuff with the casing for OpenAPI generation, whatever, that stuff shouldn't be bleeding into the godoc in a way that causes us to diverge from the godoc standards.

While I disagree with bending the Go convention to suit some specific generator pipeline's weird deficiencies or impedance mismatches with Go, I guess it's a much bigger discussion.